File: 2024/day22.go
package main

import (
	"lukechampine.com/advent/utils"
)

var input = utils.Input(2024, 22)

func next(x int) int {
	x ^= x << 6
	x &= (1 << 24) - 1
	x ^= x >> 5
	x &= (1 << 24) - 1
	x ^= x << 11
	x &= (1 << 24) - 1
	return x
}

func main() {
	secrets := utils.ExtractInts(input)
	utils.Println(utils.Sum(len(secrets), func(i int) int {
		n := secrets[i]
		for i := 0; i < 2000; i++ {
			n = next(n)
		}
		return n
	}))

	sales := make(map[[4]int8]int)
	for _, n := range secrets {
		var prevPrice int8
		var deltas [4]int8
		seen := make(map[[4]int8]bool, 2000)
		for i := 0; i <= 2000; i++ {
			price := int8(n % 10)
			n = next(n)
			copy(deltas[:], deltas[1:])
			deltas[3] = price - prevPrice
			prevPrice = price
			if i >= 4 && !seen[deltas] {
				sales[deltas] += int(price)
				seen[deltas] = true
			}
		}
	}
	var best int
	for _, v := range sales {
		best = utils.Max(best, v)
	}
	utils.Println(best)
}

File: 2024/rust/src/day22.rs
use std::collections::{HashMap, HashSet};

use crate::utils;

fn next(mut x: i64) -> i64 {
    x ^= x << 6;
    x &= (1 << 24) - 1;
    x ^= x >> 5;
    x &= (1 << 24) - 1;
    x ^= x << 11;
    x &= (1 << 24) - 1;
    x
}

pub fn part1(input: &str) -> String {
    let secrets = utils::ints(input);
    secrets
        .map(|mut n| {
            for _ in 0..2000 {
                n = next(n)
            }
            return n;
        })
        .sum::<i64>()
        .to_string()
}

pub fn part2(input: &str) -> String {
    let secrets = utils::ints(input);
    let mut sales = HashMap::new();
    for mut n in secrets {
        let mut prev_price = 0;
        let mut deltas = [0; 4];
        let mut seen = HashSet::new();
        for i in 0..=2000 {
            let price = n % 10;
            n = next(n);
            deltas.rotate_left(1);
            deltas[3] = price - prev_price;
            prev_price = price;
            if i >= 4 && !seen.contains(&deltas) {
                *sales.entry(deltas).or_insert(0) += price;
                seen.insert(deltas);
            }
        }
    }
    sales.values().max().unwrap().to_string()
}
